src/NexusMods.Library/LibraryService.cs

```diff
@@ -0,0 +1,199 @@
+
+        _downloadActivitySourceCache
+            .Connect()
+            .WhenPropertyChanged(downloadActivity => downloadActivity.Status)
+            .Where(propertyValue => propertyValue.Value == PersistedDownloadStatus.Completed)
+            .Select(propertyValue => propertyValue.Sender)
+            .SelectMany(AddCompletedDownloadAsync)
+            .Subscribe()
+            .DisposeWith(_compositeDisposable);
+    }
+
+    /// <inheritdoc/>
+    public void EnqueueDownload(IDownloadActivity downloadActivity, bool addPaused = false)
+    {
+        _logger.LogInformation("Adding download `{Title}` to the library", downloadActivity.Title);
+
+        _downloadActivitySourceCache.AddOrUpdate(downloadActivity);
+
+        if (addPaused) return;
+        downloadActivity.Downloader.Start(downloadActivity);
+    }
+
+    private async Task<LibraryFile.ReadOnly> AddCompletedDownloadAsync(
+        IDownloadActivity downloadActivity,
+        CancellationToken cancellationToken)
+    {
+        _downloadActivitySourceCache.Remove(downloadActivity);
+
+        using var tx = _connection.BeginTransaction();
+        var entityId = tx.TempId();
+
+        // TODO: create download specific library item
+        var libraryFile = await AddLibraryFile(tx, entityId, downloadActivity.DownloadPath, cancellationToken: cancellationToken);
+
+        var result = await tx.Commit();
+        return result.Remap(libraryFile);
+    }
+
+    /// <inheritdoc/>
+    public async Task<Optional<LocalFile.ReadOnly>> AddLocalFileAsync(
+        AbsolutePath absolutePath,
+        CancellationToken cancellationToken = default)
+    {
+        _logger.LogInformation("Adding local file at `{Path}` to the library", absolutePath);
+
+        if (!absolutePath.FileExists)
+        {
+            if (absolutePath.DirectoryExists())
+            {
+                _logger.LogError("File at `{Path}` can't be added to the library because it's a directory", absolutePath);
+            }
+
+            _logger.LogError("File at `{Path}` can't be added to the library because it doesn't exist", absolutePath);
+            return Optional<LocalFile.ReadOnly>.None;
+        }
+
+        using var tx = _connection.BeginTransaction();
+        var entityId = tx.TempId();
+
+        var libraryFile = await AddLibraryFile(tx, entityId, absolutePath, cancellationToken: cancellationToken);
+
+        var localFile = new LocalFile.New(tx, entityId)
+        {
+            LibraryFile = libraryFile,
+            OriginalPath = absolutePath.GetFullPath(),
+        };
+
+        var result = await tx.Commit();
+        return result.Remap(localFile);
+    }
+
+    private async Task<LibraryFile.New> AddLibraryFile(
+        ITransaction tx,
+        EntityId entityId,
+        AbsolutePath filePath,
+        CancellationToken cancellationToken)
+    {
+        var (hash, isArchive) = await AnalyzeFileAsync(filePath, cancellationToken: cancellationToken);
```

Comment:
This can take quite a while for larger files being analyzed for the first time, the old system had an `Analyzing` status on the `PersistedDownloadStatus` enum.

Do you still think it should not be in the status? If not, what do you think would be the best approach to keep track of this in the UI?

E.g. we want to show "Analizing" text and a spinner on the download entry in the UI while Analyzing work is being done.

Reply:
The download will be marked as completed once it has finished downloading. Hashing and extracting the archive are two different activities that may be linked to the "parent" download activity. I really want to separate these concepts and make the process work more like a pipeline.

The only thing the download activity is concerned with is downloading.